### ISSUE TITLE
Is it possible to produce masks on images without providing prompts?

### ISSUE THREAD

**Issue by eypros** [facebookresearch/sam2#438]
I am trying to finetune (or train from scratch as an alternative) a Sam2 model with a custom training and eval set. 

For simplicity (and to check that my model can actually overfit and, thus, work as expected) I use the same training and evaluation set (in other words I am evaluating on my training set).

I am basically following the approach of [fine-tune-train_segment_anything_2_in_60_lines_of_code](https://github.com/sagieppel/fine-tune-train_segment_anything_2_in_60_lines_of_code). 

I introduced a second dataloader (which is not really elaborated after all) for the evaluation and have split the training into epochs (I omit some samples at each epoch which do not fully complete the final batch). So, after each epoch I run the evaluation code (for getting the IoU over the whole eval set). I noticed that using the train pipeline:

            predictor.set_image_batch([image]*batch_size)
            mask_input, unnorm_coords, labels, unnorm_box = predictor._prep_prompts(input_point, input_label, box=None,
                                                                                    mask_logits=None,
                                                                                    normalize_coords=True)
            sparse_embeddings, dense_embeddings = predictor.model.sam_prompt_encoder(points=(unnorm_coords, labels),
                                                                                     boxes=None, masks=None, )

            high_res_features = [feat_level[-1].unsqueeze(0) for feat_level in predictor._features["high_res_feats"]]
            low_res_masks, prd_scores, _, _ = predictor.model.sam_mask_decoder(
                image_embeddings=predictor._features["image_embed"],
                image_pe=predictor.model.sam_prompt_encoder.get_dense_pe(), sparse_prompt_embeddings=sparse_embeddings,
                dense_prompt_embeddings=dense_embeddings, multimask_output=True, repeat_image=False,
                high_res_features=high_res_features, )
            # Upscale the masks to the original image resolution
            prd_masks = predictor._transforms.postprocess_masks(low_res_masks, predictor._orig_hw[-1])
            prd_mask = torch.sigmoid(prd_masks[:, 0]) ]()

and the code from the inference pipeline:
```
            predictor.set_image(image)
            masks, _, _ = predictor.predict(point_coords=None, box=None, multimask_output=False)

```
produced quite different results. 

a) I was wondering if this has only to do with the provided prompts (which are present in the training pipeline but not in the inference one) or there is some more differences (due to architecture etc for example).

b) I noticed that the prediction is highly affected by the provided prompts. My needs is to apply Sam2 on unprompted images and thus I can not provide any prompt. Does this mean that I cannot use Sam2 without provided prompts and I should resolve to a different model for example?

**Comment by catalys1** [comment]
SAM2 does require prompts. There is an [automatic mask generator](https://github.com/facebookresearch/sam2/blob/c2ec8e14a185632b0a5d8b161928ceb50197eddc/sam2/automatic_mask_generator.py) available, which just adds a bunch of prompts on a grid and does some post processing, if you want to look into that.

**Comment by tutu941008** [comment]
> 我正在尝试使用自定义训练和评估集微调（或从头开始训练作为替代方案）Sam2 模型。
> 
> 为了简单起见（并检查我的模型是否真的可以过拟合，从而按预期工作），我使用相同的训练集和评估集（换句话说，我在训练集上进行评估）。
> 
> 我基本上遵循了 [fine-tune-train_segment_anything_2_in_60_lines_of_code](https://github.com/sagieppel/fine-tune-train_segment_anything_2_in_60_lines_of_code) 的方法。
> 
> 我引入了第二个数据加载器（毕竟并没有真正详细说明）用于评估，并将训练分成多个 epoch（我在每个 epoch 省略了一些样本，这些样本没有完全完成最终批次）。因此，在每个 epoch 之后，我运行评估代码（用于获取整个评估集的 IoU）。我注意到使用 train 管道：
> 
> ```
>         predictor.set_image_batch([image]*batch_size)
>         mask_input, unnorm_coords, labels, unnorm_box = predictor._prep_prompts(input_point, input_label, box=None,
>                                                                                 mask_logits=None,
>                                                                                 normalize_coords=True)
>         sparse_embeddings, dense_embeddings = predictor.model.sam_prompt_encoder(points=(unnorm_coords, labels),
>                                                                                  boxes=None, masks=None, )
> 
>         high_res_features = [feat_level[-1].unsqueeze(0) for feat_level in predictor._features["high_res_feats"]]
>         low_res_masks, prd_scores, _, _ = predictor.model.sam_mask_decoder(
>             image_embeddings=predictor._features["image_embed"],
>             image_pe=predictor.model.sam_prompt_encoder.get_dense_pe(), sparse_prompt_embeddings=sparse_embeddings,
>             dense_prompt_embeddings=dense_embeddings, multimask_output=True, repeat_image=False,
>             high_res_features=high_res_features, )
>         # Upscale the masks to the original image resolution
>         prd_masks = predictor._transforms.postprocess_masks(low_res_masks, predictor._orig_hw[-1])
>         prd_mask = torch.sigmoid(prd_masks[:, 0]) ]()
> ```
> 
> 以及来自推理管道的代码：
> 
> ```
>             predictor.set_image(image)
>             masks, _, _ = predictor.predict(point_coords=None, box=None, multimask_output=False)
> ```
> 
> 产生了完全不同的结果。
> 
> a） 我想知道这是否仅与提供的提示有关（这些提示存在于训练管道中，但不存在于推理管道中），还是存在更多差异（例如，由于架构等原因）。
> 
> b） 我注意到预测受到提供的提示的很大影响。我的需求是在无提示的图像上应用 Sam2，因此我无法提供任何提示。这是否意味着在没有提供提示的情况下我无法使用 Sam2，例如我应该解析为不同的模型？

May I ask what format your dataset is in when you use your own dataset for training? Currently, I am using the semantic segmentation format, with RGB images and corresponding mask images. In the mask images, objects of the same class are assigned the same pixel value. Is this format correct? If possible, could you provide a few examples from your dataset? Thank you very much!

**Comment by yufeixue-ai** [comment]
> SAM2 does require prompts. There is an [automatic mask generator](https://github.com/facebookresearch/sam2/blob/c2ec8e14a185632b0a5d8b161928ceb50197eddc/sam2/automatic_mask_generator.py) available, which just adds a bunch of prompts on a grid and does some post processing, if you want to look into that.

Hi, I wonder if it works for the val/test datasets or not. Because I want to evaluate the performance of SAM2. Given the provided SA-V val/test dataset, there are only the video frames and mask pairs and no corresponding prompts provided.
